solidity/contracts/zeto_anon.sol

```diff
@@ -61,28 +67,66 @@ contract Zeto_Anon is IZeto, ZetoBase, ZetoFungibleWithdraw, UUPSUpgradeable {
      * Emits a {UTXOTransfer} event.
      */
     function transfer(
-        uint256[2] memory inputs,
-        uint256[2] memory outputs,
+        uint256[] memory inputs,
+        uint256[] memory outputs,
         Commonlib.Proof calldata proof,
         bytes calldata data
     ) public returns (bool) {
+        // Check and pad inputs and outputs based on the max size
+        (inputs, outputs) = checkAndPadCommitments(inputs, outputs, MAX_BATCH);
+
         require(
             validateTransactionProposal(inputs, outputs, proof),
             "Invalid transaction proposal"
         );
 
-        // construct the public inputs
-        uint256[4] memory publicInputs;
-        publicInputs[0] = inputs[0];
-        publicInputs[1] = inputs[1];
-        publicInputs[2] = outputs[0];
-        publicInputs[3] = outputs[1];
-
         // Check the proof
-        require(
-            verifier.verifyProof(proof.pA, proof.pB, proof.pC, publicInputs),
-            "Invalid proof"
-        );
+        if (inputs.length > 2) {
+            // construct the public inputs
+            uint256[20] memory publicInputs;
+            uint256 piIndex = 0;
```

Comment:
I started with that approach, using a dynamic array at the top with a piece of reusable logic. And then use the `if` `else` branch to create fixed sized array so that they can be used by the auto-generated solidity verifier code.

However, that increased the storage allocation of the smart contract. As the work that can be done using 1 fixed sized array would need 1 dynamic array + 1 fixed-size array.

Therefore, I landed on where it is right now.

Reply:
with some testing, it seems the extra cost can easily be outweighed by readability.

Added a commit to refactor the code